### ISSUE TITLE
Clicking on save twice while editing a configurable product causes current variations to disappear

### ISSUE THREAD

**Issue by samgranger** [magento/magento2#7303]
### Preconditions

1. Magento2
2. Chrome

### Steps to reproduce
1. Edit a configurable product and click on save.
2. Be impatient and click on save again a few times, just really go at it.
3. The 'current variations' are gone.

### Expected result
1. Current variations shouldn't disappear

### Actual result
1. Current variations are no longer current.

**Comment by fowbi** [comment]
We see the same behaviour in 2 of our installations, both running 2.1.1.

**Comment by fowbi** [comment]
The issue is caused when passing the configurable matrix for a second (or nth-time) to the controller via Ajax. The saveFormHandler in variations.js (in the ConfigurableProduct module) serialises the configurable-matrix value to **configurable-matrix-serialized**, but deletes the original **configurable-matrix** value afterwards. When the save button is pressed a second time, the **configurable-matrix** value will be empty and no simple products will be linked via the controller action.

Issue can be solved  by editing the saveFormHandler in variations.js (https://github.com/magento/magento2/blob/develop/app/code/Magento/ConfigurableProduct/view/adminhtml/web/js/variations/variations.js#L306)

```
          saveFormHandler: function() {
              this.source.data["product"]["configurable-matrix-serialized"] =
                  JSON.stringify(this.source.data["configurable-matrix"]);
~             //delete this.source.data["configurable-matrix"];
```

**Comment by samgranger** [comment]
Thanks for the additional information fowbi, hopefully Magento will sort this issue out with your fix.

**Comment by yaroslavGoncharuk** [comment]
@samgranger, thanks for reporting this issue.
We've created internal ticket MAGETWO-61608 to address this issue.

**Comment by Isaiaham** [comment]
@fowbi @samgranger 

For some reason, I cannot find the variations.js file under the file structure you reference above. I can go as far as html/app however, the /code folder does not exist. Where can I properly find this file in order to fix this annoying bug?

**Comment by magento-engcom-team** [comment]
@samgranger, thank you for your report.
We've created internal ticket(s) MAGETWO-69704 to track progress on the issue.

**Comment by alepane21** [comment]
It's really worst than what you could think. If the admin validation has any error (like an empty title) and you click on save, than fix the title and click on save again ... BAM!
You lose all your configurables data!
The problem has been introduced in 2.1.1 with the following commit https://github.com/magento/magento2/commit/b75fc4f83aa2fdf093ceae25449fa0b95bbc4020

If someone need a fix for Magento 2.1.10, here is a patch: https://raw.githubusercontent.com/Hevelop/magento2-patches/master/Patch-Magento_ConfigurableProduct-M2.1.10-delete-simple-products-on-validation.patch

**Comment by engcom-Echo** [comment]
Hello All

I am not able to reproduce this issue on the 2.4-develop branch by provided steps.

**Testing scenario:**

**Preconditions:**
1. Magento 2.4
2. Chrome (Version 78.0.3904.108 (Official Build) (64-bit))
3. Configurable product is created.

**Steps to reproduce:**

1. Edit a configurable product and click on save. 
2. Be impatient and click on save again a few times, just really go at it.

**Result**
Current variations are not disappear
![Peek 2019-12-09 12-38](https://user-images.githubusercontent.com/51681435/70429341-1134d580-1a81-11ea-9d95-4a22bb2c2a52.gif)

So I have to close this issue.
Thanks for your report!
